proposals/20210126-nc-and-ncp.md

```diff
@@ -0,0 +1,218 @@
+4. As a machine, I want to know how to address my control plane.
+
+### Features from user stories
+
+#### NestedCluster
+
+Kubernetes API Group: `infrastructure.cluster.x-k8s.io/v1alpha4`
+
+This resource is responsible for getting the Cluster API `Cluster` type and setting the `spec.infrasteructureRef` pointing back to to this `NestedCluster`, this allows the CAPI controllers to understand the state of the cluster. As well It listens to the `NestedControlPlane` resource which is associated and keeps it's status in sync as the cluster is provided.
+
+
+```go=
+type NestedClusterSpec struct {
+	// ControlPlaneEndpoint represents the endpoint used to communicate with the control plane.
+	// +optional
+	ControlPlaneEndpoint clusterv1.APIEndpoint `json:"controlPlaneEndpoint"`
+}
+
+type NestedClusterStatus struct {
+	// Ready is when the NestedControlPlane has a API server URL.
+	// +optional
+	Ready bool `json:"ready,omitempty"`
+}
+
+type NestedCluster struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   NestedClusterSpec   `json:"spec,omitempty"`
+	Status NestedClusterStatus `json:"status,omitempty"`
+}
+```
+
+#### NestedControlPlane
+
+Kubernetes API Group: `controlplane.cluster.x-k8s.io/v1alpha4`
+
+The NestedControlPlane resource is responsible for orchestrating the overarching cluster, this controller doesn't create any of the downstream objects instead it creates a place where the downstream objects can look up shared values. 
+
+**Question:**
+1. should we have information necessary for the syncer within this object? 
+
+```go=
+type NestedControlPlaneSpec struct {
+	// NamespacePrefix is the namespace where the control plane is deployed as well as the prefix ++ tenant control plane namespace name for all workloads 
+	// +optional
+	NamespacePrefix string `json:"namespacePrefix,omitempty"`
+    
+	// ControlPlaneEndpoint represents the endpoint used to communicate with the control plane.
+	// +optional
+	ControlPlaneEndpoint clusterv1.APIEndpoint `json:"controlPlaneEndpoint"`
+
+    
+	// EtcdRef is the reference to the NestedEtcd 
+	EtcdRef *corev1.ObjectReference `json:"etcd,omitempty"` 
+    
+	// APIServerRef is the reference to the NestedAPIServer 
+	// +optional
+	APIServerRef *corev1.ObjectReference `json:"apiserver,omitempty"` 
+    
+	// ContollerManagerRef is the reference to the NestedControllerManager
+	// +optional
+	ControllerManagerRef *corev1.ObjectReference `json:"controllerManager,omitempty"`
+}
+
+type NestedControlPlaneStatus struct {
+	// ExternalManagedControlPlane indicates to cluster-api that the control plane
+	// is managed by an external service such as AKS, EKS, GKE, etc.
+	// +kubebuilder:default=true
+	ExternalManagedControlPlane *bool `json:"externalManagedControlPlane,omitempty"`
+    
+	// Initialized denotes whether or not the control plane has the
+	// uploaded kubernetes config-map.
```

Comment:
I am not quite sure what does the `uploaded kubernetes config-map` means.

Reply:
This is a little specific to KubeadmControlPlane and how a kubeadm cluster will upload the config used to generate the cluster into the `kube-system` namespace of the tenant control plane.